Added code snippets of C++ on bit manipulation, math & number, searching, sorting

ashukr07 opened the pull request:
<!-- **ANY PULL REQUEST NOT FOLLOWING GUIDELINES OR NOT INCLUDING A DESCRIPTION WILL BE CLOSED !** -->

# Description
Added code snippets for **C++**:

1. **Math and Number Category** snippets:
    - Binary to decimal conversion
    - Check if the number is perfect number
    - Calculate the factorial of a number using recursion
    - Find the nth fibonacci number using recursion
    - Calculate greatest comman divisor(GCD)
    - Calculate the least comman multiple(LCM)
    - Calculate the sum of digits of a number

2. Added new **Bit Manipulation Category**
    - Check if a given number is power of 2 
    - Count set bits in a number
    - Find the non repeating number in array
    - Reverse bits
    - Calculate the XOR over 1 to n(provided by user)

3. Added new **Searching Category**
    - Linear Search
    - Binary Search

4. Added new **Sorting Category**
    - Bubble sort
    - Insertion sort
    - Merge sort
    - Quick sort
    - Selection sort

## Type of Change

<!-- What kind of change does this pull request introduce? (Check all that apply) -->

- [x] ✨ New snippet
- [ ] 🛠 Improvement to an existing snippet
- [ ] 🐞 Bug fix
- [ ] 📖 Documentation update
- [ ] 🔧 Other (please describe):

## Checklist

<!--  Before submitting, ensure your pull request meets these requirements: -->

- [x] I have tested my code and verified it works as expected.
- [x] My code follows the style and contribution guidelines of this project.
- [x] Comments are added where necessary for clarity.
- [ ] Documentation has been updated (if applicable).
- [x] There are no new warnings or errors from my changes.

## Related Issues

<!-- Link any relevant issues (use #issue-number syntax). If not, leave it empty -->

Closes #

## Additional Context

<!-- Add any extra details, questions, or considerations here. -->

## Screenshots (Optional)

<!-- If your changes affect visuals, please include screenshots. -->

<details> 
<summary>Click to view screenshots</summary> 

<!-- Add your screenshots here --> 

</details>

saminjay commented on snippets/cpp/debuging/vector-print.md:
Can you clarify the reason for deleting this file?

ashukr07 replied:
There were 2 folders named **debugging** and **debuging** and both were having same snippets.
So, to avoid redundancy I removed debuging folder.

saminjay replied:
Right, didn't realize. Thanks

majvax commented:
Hey, I understand that bit manipulation isn't exclusive to C and can also be done in C++. However, since the snippets don't use any C++-specific features (at least no big one) and are more aligned with C-style programming, it makes sense to categorize them under C.

ashukr07 commented:
> Hey, I understand that bit manipulation isn't exclusive to C and can also be done in C++. However, since the snippets don't use any C++-specific features (at least no big one) and are more aligned with C-style programming, it makes sense to categorize them under C.

Got it, thanks for pointing out. I’ll make the changes to categorize the snippets under C accordingly.

ashukr07 commented:
I have moved all the snippets of bit manipulation to C except the find the non repeating element in vector as vector is supported only in C++. 
I have also removed the searching and sorting snippets and gcd and lcm snippets as it is present in C++ stl.

Mathys-Gasnier commented on snippets/cpp/math-and-numbers/sum-of-digits.md:
This does not use C++ features, should be moved to C

Mathys-Gasnier commented on snippets/cpp/math-and-numbers/fibonacci-number.md:
This does not use C++ features, should be moved to C

Mathys-Gasnier commented on snippets/cpp/math-and-numbers/factorial.md:
This does not use C++ features, should be removed since it already exists in the C language

Mathys-Gasnier commented on snippets/cpp/math-and-numbers/compound-interest.md:
This does not use C++ features, should be moved to C

Mathys-Gasnier commented on snippets/cpp/math-and-numbers/check-perfect-number.md:
This does not use C++ features, should be moved to C

Mathys-Gasnier reviewed (changes requested):
Some little misplacements of files, but overall great work

ashukr07 commented:
All the mentioned snippets moved to C

Mathys-Gasnier reviewed (approved):
Great work thanks !
